// boardgame.io-angular/esm2015/lib/bio-debug.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import { Debug } from 'boardgame.io/dist/react';
import * as uuid from 'uuid';
import { Component, Input } from '@angular/core';
/**
 * @record
 */
export function GameState() { }
function GameState_tsickle_Closure_declarations() {
    /** @type {?} */
    GameState.prototype.G;
    /** @type {?} */
    GameState.prototype.ctx;
    /** @type {?} */
    GameState.prototype.log;
    /** @type {?|undefined} */
    GameState.prototype.isActive;
    /** @type {?} */
    GameState.prototype._initial;
}
/**
 * @record
 */
export function DebugProps() { }
function DebugProps_tsickle_Closure_declarations() {
    /** @type {?} */
    DebugProps.prototype.gamestate;
    /** @type {?} */
    DebugProps.prototype.gameID;
    /** @type {?|undefined} */
    DebugProps.prototype.playerID;
    /** @type {?|undefined} */
    DebugProps.prototype.isConnected;
    /** @type {?|undefined} */
    DebugProps.prototype.isMultiplayer;
    /** @type {?|undefined} */
    DebugProps.prototype.moves;
    /** @type {?|undefined} */
    DebugProps.prototype.events;
    /** @type {?|undefined} */
    DebugProps.prototype.restore;
    /** @type {?|undefined} */
    DebugProps.prototype.showLog;
    /** @type {?|undefined} */
    DebugProps.prototype.store;
    /** @type {?|undefined} */
    DebugProps.prototype.step;
    /** @type {?|undefined} */
    DebugProps.prototype.reset;
    /** @type {?|undefined} */
    DebugProps.prototype.reducer;
    /** @type {?|undefined} */
    DebugProps.prototype.overrideGameState;
    /** @type {?|undefined} */
    DebugProps.prototype.visualizeAI;
}
export class DebugComponent {
    /**
     * @return {?}
     */
    getRootDomNode() {
        return document.getElementById(this.rootDomID);
    }
    /**
     * @return {?}
     */
    getProps() {
        const { gamestate, gameID, playerID, isConnected, isMultiplayer, moves, events, restore, showLog, store, step, reset, reducer, overrideGameState, visualizeAI, } = this;
        return {
            gamestate,
            gameID,
            playerID,
            isConnected,
            isMultiplayer,
            moves,
            events,
            restore,
            showLog,
            store,
            step,
            reset,
            reducer,
            overrideGameState,
            visualizeAI,
        };
    }
    /**
     * @return {?}
     */
    isMounted() {
        return !!this.rootDomID;
    }
    /**
     * @return {?}
     */
    render() {
        if (this.isMounted()) {
            ReactDOM.render(React.createElement(Debug, this.getProps()), this.getRootDomNode());
        }
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.rootDomID = uuid.v1();
    }
    /**
     * @return {?}
     */
    ngOnChanges() {
        this.render();
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.render();
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        ReactDOM.unmountComponentAtNode(this.getRootDomNode());
    }
}
DebugComponent.decorators = [
    { type: Component, args: [{
                selector: 'bio-debug',
                template: '<span [id]="rootDomID"></span>'
            },] },
];
DebugComponent.propDecorators = {
    gamestate: [{ type: Input }],
    gameID: [{ type: Input }],
    playerID: [{ type: Input }],
    isConnected: [{ type: Input }],
    isMultiplayer: [{ type: Input }],
    moves: [{ type: Input }],
    events: [{ type: Input }],
    restore: [{ type: Input }],
    showLog: [{ type: Input }],
    store: [{ type: Input }],
    step: [{ type: Input }],
    reset: [{ type: Input }],
    reducer: [{ type: Input }],
    overrideGameState: [{ type: Input }],
    visualizeAI: [{ type: Input }]
};
function DebugComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    DebugComponent.prototype.gamestate;
    /** @type {?} */
    DebugComponent.prototype.gameID;
    /** @type {?} */
    DebugComponent.prototype.playerID;
    /** @type {?} */
    DebugComponent.prototype.isConnected;
    /** @type {?} */
    DebugComponent.prototype.isMultiplayer;
    /** @type {?} */
    DebugComponent.prototype.moves;
    /** @type {?} */
    DebugComponent.prototype.events;
    /** @type {?} */
    DebugComponent.prototype.restore;
    /** @type {?} */
    DebugComponent.prototype.showLog;
    /** @type {?} */
    DebugComponent.prototype.store;
    /** @type {?} */
    DebugComponent.prototype.step;
    /** @type {?} */
    DebugComponent.prototype.reset;
    /** @type {?} */
    DebugComponent.prototype.reducer;
    /** @type {?} */
    DebugComponent.prototype.overrideGameState;
    /** @type {?} */
    DebugComponent.prototype.visualizeAI;
    /** @type {?} */
    DebugComponent.prototype.rootDomID;
}
