// boardgame.io-angular/esm2015/lib/bio-client.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { Component, Input } from '@angular/core';
import { Client as RawClient } from 'boardgame.io/dist/client';
export class ClientComponent {
    constructor() {
        // instance inputs
        this.gameID = 'default';
        this.playerID = null;
        this.credentials = null;
        // overridable inputs (both class and instance)
        this.debug = true;
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        this.client = RawClient({
            game: this.game,
            ai: this.ai,
            numPlayers: this.numPlayers,
            multiplayer: this.multiplayer,
            gameID: this.gameID,
            playerID: this.playerID,
            credentials: this.credentials,
            enhancer: this.enhancer,
        });
        // todo is this needed? (forceUpdate on react)
        // this.client.subscribe(() => {
        //   console.log('client.subscribe triggered (forceUpdate needed?)');
        // });
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        if (this.client) {
            // noinspection TsLint
            if (changes["gameID"] && changes["gameID"].currentValue != changes["gameID"].previousValue) {
                this.client.updateGameID(changes["gameID"].currentValue);
            }
            // noinspection TsLint
            if (changes["playerID"] && changes["playerID"].currentValue != changes["playerID"].previousValue) {
                this.client.updatePlayerID(changes["playerID"].currentValue);
            }
            // noinspection TsLint
            if (changes["credentials"] && changes["credentials"].currentValue != changes["credentials"].previousValue) {
                this.client.updateCredentials(changes["credentials"].currentValue);
            }
        }
    }
    /**
     * @return {?}
     */
    ngAfterContentInit() {
        this.client.connect();
    }
    /**
     * @return {?}
     */
    getBoardInputs() {
        const /** @type {?} */ state = this.client.getState();
        return Object.assign({}, state, { isMultiplayer: this.multiplayer !== undefined, moves: this.client.moves, events: this.client.events, gameID: this.gameID, playerID: this.playerID, reset: this.client.reset, undo: this.client.undo, redo: this.client.redo });
    }
}
ClientComponent.decorators = [
    { type: Component, args: [{
                selector: 'bio-client',
                template: `
    <ng-container *ngIf="client">
      <ng-container *ngComponentOutlet="board;
                                            ndcDynamicInputs: getBoardInputs()">
      </ng-container>
    </ng-container>

    <bio-debug *ngIf="debug && client"
               [gamestate]="client.getState()"
               [gameID]="gameID"
               [playerID]="playerID"
               [isMultiplayer]="multiplayer !== undefined"
               [moves]="client.moves"
               [events]="client.events"
               [store]="client.store"
               [step]="client.step"
               [reset]="client.reset"
               [reducer]="client.reducer"
    ></bio-debug>
  `
            },] },
];
ClientComponent.propDecorators = {
    game: [{ type: Input }],
    numPlayers: [{ type: Input }],
    board: [{ type: Input }],
    multiplayer: [{ type: Input }],
    ai: [{ type: Input }],
    enhancer: [{ type: Input }],
    gameID: [{ type: Input }],
    playerID: [{ type: Input }],
    credentials: [{ type: Input }],
    debug: [{ type: Input }]
};
function ClientComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ClientComponent.prototype.game;
    /** @type {?} */
    ClientComponent.prototype.numPlayers;
    /** @type {?} */
    ClientComponent.prototype.board;
    /** @type {?} */
    ClientComponent.prototype.multiplayer;
    /** @type {?} */
    ClientComponent.prototype.ai;
    /** @type {?} */
    ClientComponent.prototype.enhancer;
    /** @type {?} */
    ClientComponent.prototype.gameID;
    /** @type {?} */
    ClientComponent.prototype.playerID;
    /** @type {?} */
    ClientComponent.prototype.credentials;
    /** @type {?} */
    ClientComponent.prototype.debug;
    /** @type {?} */
    ClientComponent.prototype.client;
}
