// boardgame.io-angular/esm5/lib/bio-client.component.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import * as tslib_1 from "tslib";
import { Component, Input } from '@angular/core';
import { Client as RawClient } from 'boardgame.io/dist/client';
var ClientComponent = /** @class */ (function () {
    function ClientComponent() {
        // instance inputs
        this.gameID = 'default';
        this.playerID = null;
        this.credentials = null;
        // overridable inputs (both class and instance)
        this.debug = true;
    }
    /**
     * @return {?}
     */
    ClientComponent.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.client = RawClient({
            game: this.game,
            ai: this.ai,
            numPlayers: this.numPlayers,
            multiplayer: this.multiplayer,
            gameID: this.gameID,
            playerID: this.playerID,
            credentials: this.credentials,
            enhancer: this.enhancer,
        });
        // todo is this needed? (forceUpdate on react)
        // this.client.subscribe(() => {
        //   console.log('client.subscribe triggered (forceUpdate needed?)');
        // });
    };
    /**
     * @param {?} changes
     * @return {?}
     */
    ClientComponent.prototype.ngOnChanges = /**
     * @param {?} changes
     * @return {?}
     */
    function (changes) {
        if (this.client) {
            // noinspection TsLint
            if (changes["gameID"] && changes["gameID"].currentValue != changes["gameID"].previousValue) {
                this.client.updateGameID(changes["gameID"].currentValue);
            }
            // noinspection TsLint
            if (changes["playerID"] && changes["playerID"].currentValue != changes["playerID"].previousValue) {
                this.client.updatePlayerID(changes["playerID"].currentValue);
            }
            // noinspection TsLint
            if (changes["credentials"] && changes["credentials"].currentValue != changes["credentials"].previousValue) {
                this.client.updateCredentials(changes["credentials"].currentValue);
            }
        }
    };
    /**
     * @return {?}
     */
    ClientComponent.prototype.ngAfterContentInit = /**
     * @return {?}
     */
    function () {
        this.client.connect();
    };
    /**
     * @return {?}
     */
    ClientComponent.prototype.getBoardInputs = /**
     * @return {?}
     */
    function () {
        var /** @type {?} */ state = this.client.getState();
        return tslib_1.__assign({}, state, { isMultiplayer: this.multiplayer !== undefined, moves: this.client.moves, events: this.client.events, gameID: this.gameID, playerID: this.playerID, reset: this.client.reset, undo: this.client.undo, redo: this.client.redo });
    };
    ClientComponent.decorators = [
        { type: Component, args: [{
                    selector: 'bio-client',
                    template: "\n    <ng-container *ngIf=\"client\">\n      <ng-container *ngComponentOutlet=\"board;\n                                            ndcDynamicInputs: getBoardInputs()\">\n      </ng-container>\n    </ng-container>\n\n    <bio-debug *ngIf=\"debug && client\"\n               [gamestate]=\"client.getState()\"\n               [gameID]=\"gameID\"\n               [playerID]=\"playerID\"\n               [isMultiplayer]=\"multiplayer !== undefined\"\n               [moves]=\"client.moves\"\n               [events]=\"client.events\"\n               [store]=\"client.store\"\n               [step]=\"client.step\"\n               [reset]=\"client.reset\"\n               [reducer]=\"client.reducer\"\n    ></bio-debug>\n  "
                },] },
    ];
    ClientComponent.propDecorators = {
        game: [{ type: Input }],
        numPlayers: [{ type: Input }],
        board: [{ type: Input }],
        multiplayer: [{ type: Input }],
        ai: [{ type: Input }],
        enhancer: [{ type: Input }],
        gameID: [{ type: Input }],
        playerID: [{ type: Input }],
        credentials: [{ type: Input }],
        debug: [{ type: Input }]
    };
    return ClientComponent;
}());
export { ClientComponent };
function ClientComponent_tsickle_Closure_declarations() {
    /** @type {?} */
    ClientComponent.prototype.game;
    /** @type {?} */
    ClientComponent.prototype.numPlayers;
    /** @type {?} */
    ClientComponent.prototype.board;
    /** @type {?} */
    ClientComponent.prototype.multiplayer;
    /** @type {?} */
    ClientComponent.prototype.ai;
    /** @type {?} */
    ClientComponent.prototype.enhancer;
    /** @type {?} */
    ClientComponent.prototype.gameID;
    /** @type {?} */
    ClientComponent.prototype.playerID;
    /** @type {?} */
    ClientComponent.prototype.credentials;
    /** @type {?} */
    ClientComponent.prototype.debug;
    /** @type {?} */
    ClientComponent.prototype.client;
}
